diff -up openssl-fips-0.9.8e/crypto/perlasm/x86_64-xlate.pl.sign openssl-fips-0.9.8e/crypto/perlasm/x86_64-xlate.pl
--- openssl-fips-0.9.8e/crypto/perlasm/x86_64-xlate.pl.sign	2010-04-16 18:52:58.000000000 +0200
+++ openssl-fips-0.9.8e/crypto/perlasm/x86_64-xlate.pl	2010-04-16 18:53:11.000000000 +0200
@@ -186,8 +186,10 @@ my $current_function;
 	if (!$masm) {
 	    # Solaris /usr/ccs/bin/as can't handle multiplications
 	    # in $self->{label}
+	    use integer;
 	    $self->{label} =~ s/(?<![0-9a-f])(0[x0-9a-f]+)/oct($1)/egi;
 	    $self->{label} =~ s/([0-9]+\s*[\*\/\%]\s*[0-9]+)/eval($1)/eg;
+	    $self->{label} =~ s/([0-9]+)/$1<<32>>32/eg;
 
 	    if (defined($self->{index})) {
 		sprintf "%s(%%%s,%%%s,%d)",
